transactions.py

```diff
@@ -0,0 +1,222 @@
+from flask import Blueprint, request, jsonify
+from db import get_db_connection
+from sqlalchemy import text
+from middleware import jwt_required_with_user_id
+from datetime import datetime
+import uuid
+
+transactions_bp = Blueprint('transactions', __name__)
+
+@transactions_bp.route('/account/<account_id>/deposit', methods=['POST'])
+@jwt_required_with_user_id
+def deposit(current_user_id, account_id):
+    data = request.get_json()
+    amount = data.get('amount')
+    
+    with get_db_connection() as connection:
+        try:
+            # Check account ownership
+            account = connection.execute(
+                text('SELECT * FROM accounts WHERE id = :account_id AND owner_id = :user_id'),
+                {'account_id': account_id, 'user_id': current_user_id}
+            ).fetchone()
+            
+            if not account:
+                return jsonify({"error": "Account not found or unauthorized"}), 403
+            
+            # Start transaction
+            connection.begin()
+            
+            # Update balance
+            connection.execute(
+                text('UPDATE accounts SET balance_cents = balance_cents + :amount WHERE id = :account_id'),
+                {'amount': amount, 'account_id': account_id}
+            )
+            
+            # Create transaction record
+            transaction_id = str(uuid.uuid4())
+            connection.execute(
+                text("""
+                    INSERT INTO transactions 
+                    (id, initiator_id, to_bank_account_id, amount, created_at, updated_at)
+                    VALUES (:id, :initiator_id, :account_id, :amount, :created_at, :updated_at)
+                """),
+                {
+                    'id': transaction_id,
+                    'initiator_id': current_user_id,
+                    'account_id': account_id,
+                    'amount': amount,
+                    'created_at': datetime.now(),
+                    'updated_at': datetime.now()
+                }
+            )
+            
+            connection.commit()
+        except Exception as e:
+            connection.rollback()
+            return jsonify({"error": str(e)}), 500
+    
+    return jsonify({"message": "Deposit successful", "transaction_id": transaction_id}), 200
+
+@transactions_bp.route('/account/<account_id>/withdraw', methods=['POST'])
+@jwt_required_with_user_id
+def withdraw(current_user_id, account_id):
+    data = request.get_json()
+    amount = data.get('amount')
+    
+    with get_db_connection() as connection:
+        try:
+            # Check account ownership and balance
+            account = connection.execute(

```

Comment:
First thing, you want to move this one to within the transaction below, so you can "enjoy" the isolation level provided by the transaction. Keep in mind that the default isolation level for a PG transaction is READ_COMMITED so technically, in case of a concurrent requests, you may still "get" the old balance before the other transaction committed the change.
On the other hand, if you change the isolation level of this request to a lower one (e.g. READ_UNCOMMITTED), then you still have a problem that in case the other transaction rolled back, you still denied the withdrawal although you have enough balance.
If we upgrade the isolation level to "SERIALIZABLE" then you won't have this issue, BUT in case of concurrent transactions, one will fail, causing the API call to fail with a 500, and you (or the client) need to take care of retries.

Instead, the right way is to use the row-locking functionality and tell the database that once you read the value, you want to lock the row for any updates outside of the current transaction until the transaction is committed. This will prevent any race conditions on the data as the other transaction will be waiting (unless time-out).

```
SELECT * FROM accounts WHERE id = :account_id AND owner_id = :user_id FOR UPDATE
```